Feat/dns verification

sumitnitsurat opened the pull request:
- DNS verification added for document store
- Added unit test for dns verification 
- Create New issue for tls library used in dnsprove

sumitnitsurat committed:
abc

sumitnitsurat committed:
dns verification added

sumitnitsurat committed:
Merge branch 'master' into feat/dns-verification

sumitnitsurat committed:
build fix

sumitnitsurat committed:
dnsproce version bump

sumitnitsurat committed:
Merge branch 'master' into feat/dns-verification

sumitnitsurat committed:
regitry and dns verification added

sumitnitsurat committed:
spelling fixed

sumitnitsurat committed:
unit test added

- updated exiting unit test for issuer verified through registry
- added unit test fir dns verified

sumitnitsurat committed:
Merge branch 'master' into feat/dns-verification

rjchow commented on package.json:
```diff
@@ -52,6 +53,7 @@
     "redux": "^4.0.0",
     "redux-devtools-extension": "^2.13.7",
     "redux-saga": "^0.16.0",
+    "tls": "0.0.1",
```
what's this library used for?

yehjxraymond replied:
If it's suppose to be in dnsprove, shoudl move there instead.

yehjxraymond committed:
adde comments

sumitnitsurat committed:
dns verification logic changed

- Added one by one verifrication for each issuer for registry and dns.
- Modified message shown when issuer is verified

sumitnitsurat committed:
unit test for dns verification modified

sumitnitsurat committed:
Merge branch 'master' into feat/dns-verification

sumitnitsurat committed:
certificate saga test modifed

yehjxraymond commented on package.json:
```diff
@@ -40,6 +40,7 @@
     "next-seo": "^1.11.2",
     "node-sass": "^4.12.0",
     "number-to-words": "^1.2.4",
+    "opencerts-dnsprove": "^2.0.0",
```
Needs to be in @govtechsg org.

sumitnitsurat committed:
issuer verification summary modfied

- PR changes

sumitnitsurat committed:
unit test modified

sumitnitsurat committed:
dns domain name converted to uppercase

- code cleanup

sumitnitsurat committed:
code clean up

sumitnitsurat committed:
code restructured

sumitnitsurat committed:
dns string netid bug fixed

yehjxraymond committed:
Fixed some naming and functions

yehjxraymond committed:
Fixed org name for package

yehjxraymond committed:
Updated dnsprove to use fetch instead of dohdec

Nebulis commented on src/components/CertificateVerifyBlock/CertificateVerifyBlock.js:
```diff
@@ -56,7 +57,25 @@ const renderIcon = status => {
+  const verifiedText = identityStatus.reduce(
+    (prev, next) => {
+      /* eslint-disable-next-line no-param-reassign */
+      if (next.registry && !prev.isRegistry) prev.isRegistry = true;
```
minor comment but the condition `!prev.isRegistry` is not necessary

Nebulis replied:
up to the opener to resolve discussion :)

sumitnitsurat replied:
okay :p

Nebulis replied:
What's the issue on setting isRegistry again and again ?

Anyway I actually have another suggestion to change the function:

```
if(some(identityStatus, (idontknowhowtonamethis) => idontknowhowtonamethis.isRegistry)) {
  return "Accredited by SSG";
} else {
// note filter Boolean is to remove empty values
  return `Issued by ${identityStatus.map((idontknowhowtonamethis) =>idontknowhowtonamethis.dns).filter(Boolean).dns.join(" & ")}`
}
```

the issue with the current function is that there is `no comments/explanation` and the intent using a reducer is not really clear as it's doing 2 things at the same time

sumitnitsurat replied:
We are already iterating on the same array twice in our code ..with this method in case of dns it will iterate 4 times on the same array to return us the required value..

Nebulis replied:
What's the issue with that ? performance concern to iterate over tiny array ?

readabiliy > performance unless performance is an actual issue. On top of that this function iterate only twice in that case

sumitnitsurat replied:
ok

sumitnitsurat replied:
fixed it ..please check

Nebulis commented on src/components/CertificateVerifyBlock/CertificateVerifyBlock.test.js:
```diff
@@ -0,0 +1,30 @@
+    expect(getIdentityVerificationText(testValue)).toEqual("Issued by ABC.COM");
+  });
+
+  test("when multiple registry is verified", () => {
```
let's differentiate the test case in their description

Nebulis replied:
test description must reflect what the test is ding, there are the same description twice

sumitnitsurat replied:
something like this "should return appropriate display text when multiple registry is verified"

Nebulis replied:
yup this sounds better

sumitnitsurat replied:
updated please check .

Nebulis commented on src/components/CertificateVerifyBlock/CertificateVerifyBlock.test.js:
```diff
@@ -0,0 +1,30 @@
+    expect(getIdentityVerificationText(testValue)).toEqual(
+      "Issued by ABC.COM & DEMO.COM"
+    );
+  });
```
What about those use cases:
- testValue = []
- testValue = [{ registry: "demo", dns: "demo.com" }, { registry: false, dns: "abc.com" }]
- testValue = [{ registry: false, dns: "abc.com" }, { registry: "demo", dns: "demo.com" }]

Nebulis replied:
and so ?it's still possible to call that function that way

sumitnitsurat replied:
when there is no registry and dns saga test will fail. this component code will never execute.

sumitnitsurat replied:
Should I resolve this ?

sumitnitsurat committed:
PR changes

- add check to throw error if no issuers found

sumitnitsurat committed:
tets modified

sumitnitsurat committed:
unit test description updated

sumitnitsurat committed:
Merge branch 'master' into feat/dns-verification

sumitnitsurat committed:
key updated address => addr

sumitnitsurat committed:
Merge branch 'feat/dns-verification' of https://github.com/OpenCerts/opencerts-website into feat/dns-verification

rjchow reviewed (changes requested):
the verification logic looks correct but the flow is hard to follow - we'll need to refactor this in the following days.

aside from that there's also the rendering issue which i posted a screenshot of on telegram

do add in the integration test cases below as well for the various combinations

### Single Issuer
##### Valid (Green)
######Schema 1.5:
Accredited Issuer

###### Schema 2.0
Accredited Issuer
DNS Issuer

##### Invalid (means it should show red and not progress)
###### Schema 1.5
Issuer not on registry
###### Schema 2.0
No identityProof, issuer not on registry
Has identityProof, DNS records has wrong address
Has identityProof, no OpenAtts records

### Multiple Issuer
##### Valid
All accredited issuers: should show accredited cert
All DNS issuers: should show issued by DNS names
Mix of accredited issuers & DNS issuers: should show accredited cert

##### Invalid (show red and not render cert)
Mix of valid and invalid issuers
All invalid issuers

sumitnitsurat committed:
UI fixed

sumitnitsurat committed:
log fixed

sumitnitsurat commented on src/sagas/certificate.js:
```diff
@@ -253,94 +247,144 @@ function isApprovedENSDomain(issuerAddress) {
+
+  const dnsRecords = yield call(getDocumentStoreRecords, location);
+
+  trace(`DNS records: ${dnsRecords}`); // dnsRecords: [{addr: "0xabc", netId: 3}]
```
fixed it.

sumitnitsurat commented on src/sagas/certificate.test.js:
```diff
@@ -77,37 +94,42 @@ function whenThereIsOnlyOneEnsName() {
+    .addDataField("issuers", [
+      {
+        identityProof: { type: "DNS-TXT", location: dnsName },
+        documentSotre: ethereumAddress
```
fixed it.

sumitnitsurat commented:
> the verification logic looks correct but the flow is hard to follow - we'll need to refactor this in the following days.
> 
> aside from that there's also the rendering issue which i posted a screenshot of on telegram
> 
> do add in the integration test cases below as well for the various combinations
> 
> ### Single Issuer
> ##### Valid (Green)
> ######Schema 1.5:
> Accredited Issuer
> 
> ###### Schema 2.0
> Accredited Issuer
> DNS Issuer
> 
> ##### Invalid (means it should show red and not progress)
> ###### Schema 1.5
> Issuer not on registry
> 
> ###### Schema 2.0
> No identityProof, issuer not on registry
> Has identityProof, DNS records has wrong address
> Has identityProof, no OpenAtts records
> 
> ### Multiple Issuer
> ##### Valid
> All accredited issuers: should show accredited cert
> All DNS issuers: should show issued by DNS names
> Mix of accredited issuers & DNS issuers: should show accredited cert
> 
> ##### Invalid (show red and not render cert)
> Mix of valid and invalid issuers
> All invalid issuers

I will create a new issue for that to refactor and add integration test for this.

rjchow commented:
do the tests in this merge so that we can have confidence that the verification logic works - right now we don't run any tests to check that cert verification is correct so that's very dangerous

sumitnitsurat committed:
integration test added for different scenario of issuer verification (#388)

sumitnitsurat committed:
Merge branch 'master' into feat/dns-verification

sumitnitsurat committed:
dnsprove version updated

rjchow reviewed (commented):
left two comments the rest are correct

sumitnitsurat committed:
fix error view when there is unverified issuers (#389)

* fix error view when there is unverified issuers
